src/main/java/randoop/resource/ClassAnnotationScanner.java

```diff
@@ -0,0 +1,156 @@
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import org.checkerframework.checker.signature.qual.BinaryName;
+import org.checkerframework.checker.signature.qual.InternalForm;
+import org.objectweb.asm.AnnotationVisitor;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.TypePath;
+import org.objectweb.asm.TypeReference;
+import org.plumelib.reflection.Signatures;
+
+/**
+ * ClassAnnotationScanner is a ClassVisitor that collects methods that are annotated with specific
+ * annotations.
+ */
+public class ClassAnnotationScanner extends ClassVisitor {
+
+  /** The desired annotations: we will return methods that have one of these annotations. */
+  private final Collection<String> desiredAnnotations;
+
+  /** The binary name of the class being visited. */
+  private @BinaryName String className = null;
+
+  /** The accumulated set of captured fully qualified method signatures. */
+  private final Set<String> matchingSignatures = new HashSet<>();
+
+  /**
+   * Creates a new AnnotationScanner to look for methods with the specified annotations.
+   *
+   * @param api API version for ASM
+   * @param desiredAnnotations annotations to look for
+   */
+  public ClassAnnotationScanner(int api, Collection<String> desiredAnnotations) {
+    super(api);
+    this.desiredAnnotations = desiredAnnotations;
+  }
+
+  /**
+   * Returns the Randoop fully qualified signature of a method. e.g. "java.util.String.hashCode()"
+   *
+   * @param method method name
+   * @param argumentSignature signature of the method's arguments, e.g.
+   *     "(char[],int,int,java.lang.String,int)"
+   * @return Randoop fully qualified signature
+   */
+  private String getFullyQualifiedMethodSignature(String method, String argumentSignature) {
+    return className + "." + method + argumentSignature;
+  }
+
+  /**
+   * Adds a method to the set of captured methods if the given annotation is in {@code
+   * desiredAnnotations}.
+   *
+   * @param annotationName annotation name as read by ASM, e.g.
+   *     "org.checkerframework.checker.determinism.qual.NonDet"
+   * @param method method name in Randoop fully-qualified signature format.
+   * @param argumentSignature signature of the method's arguments, e.g.
+   *     "(char[],int,int,java.lang.String,int)"
+   */
+  private void addMethodIfMatching(String annotationName, String method, String argumentSignature) {
+    if (desiredAnnotations.contains(Signatures.fieldDescriptorToBinaryName(annotationName))) {
+      matchingSignatures.add(getFullyQualifiedMethodSignature(method, argumentSignature));
+    }
+  }
+
+  @Override
+  public void visit(
+      int version,
+      int access,
+      String name,
+      String signature,
+      String superName,
+      String[] interfaces) {
+
+    @SuppressWarnings("signature")
+    @InternalForm String internalFormName = name;
```

Comment:
I suggested putting the annotation on the formal parameter.  (I think that is clearer, and it avoids the need for a temporary variable.)  Can you explain why you moved it here rather than doing that?